.github/workflows/ci_linux_x64_asan-libshortfin.yml

```diff
@@ -0,0 +1,151 @@
+
+    - name: Install dependencies
+      if: steps.cache-python-asan.outputs.cache-hit != 'true'
+      run: |
+        sudo apt update
+        sudo apt install clang lld cmake ninja-build
+        sudo apt install build-essential libssl-dev zlib1g-dev libbz2-dev libreadline-dev libsqlite3-dev curl git libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev
+
+    - name: Checkout pyenv
+      if: steps.cache-python-asan.outputs.cache-hit != 'true'
+      uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
+      with:
+        repository: pyenv/pyenv
+        ref: ${{ env.PYENV_REF }}
+        path: ${{ env.PYENV_DIR }}
+
+    - name: Install pyenv & Python
+      if: steps.cache-python-asan.outputs.cache-hit != 'true'
+      run: |
+        cd ${{ env.PYENV_DIR }}
+        src/configure && make -C src
+        export PYENV_ROOT=${{ env.PYENV_DIR }}
+        export PATH=${{ env.PYENV_DIR }}/bin:$PATH && eval "$(pyenv init -)"
+        CC=clang-18 CXX=clang++-18 LDFLAGS="-lstdc++" PYTHON_CONFIGURE_OPTS="--with-address-sanitizer" pyenv install -v -g ${{ env.PYTHON_VER }}
+
+
+  build-and-test:
+    name: Build and test libshortfin
+    needs: [setup-python-asan]
+    runs-on: ubuntu-24.04
+
+    steps:
+    - name: Install dependencies
+      run: |
+        sudo apt update
+        sudo apt install clang lld cmake ninja-build
+
+    - name: Checkout repository
+      uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
+      with:
+        submodules: false
+
+    - name: Checkout IREE repo
+      uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
+      with:
+        repository: iree-org/iree
+        path: ${{ env.IREE_SOURCE_DIR }}
+        submodules: false
+
+    - name: Initalize IREE submodules
+      run : |
+        cd ${{ env.IREE_SOURCE_DIR }}
+        git submodule update --init --depth 1 -- third_party/benchmark
+        git submodule update --init --depth 1 -- third_party/cpuinfo/
+        git submodule update --init --depth 1 -- third_party/flatcc
+        git submodule update --init --depth 1 -- third_party/googletest
+        git submodule update --init --depth 1 -- third_party/hip-build-deps/
+
+    - name: Setup python
+      id: cache-python-asan
+      uses: actions/cache/restore@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
+      with:
+        path: ${{ env.PYENV_DIR }}
+        key: ${{ runner.os }}-python-asan-${{ env.PYENV_REF }}-${{ env.PYTHON_VER }}-${{ env.CACHE_VER }}
+    - name: Install Python packages
+      # TODO: Switch to `pip install -r requirements.txt -e libshortfin/`.
+      run: |
+        export PYENV_ROOT=${{ env.PYENV_DIR }}
+        export PATH=${{ env.PYENV_DIR }}/bin:$PATH && eval "$(pyenv init -)"
+        pyenv versions
+        pyenv global ${{ env.PYTHON_VER }}-debug
+        pyenv local ${{ env.PYTHON_VER }}-debug
+        pyenv shell ${{ env.PYTHON_VER }}-debug
+        pip install nanobind
```

Comment:
You should add this to requirements.txt and pin it (to 2.0). This can/has broken before and will poison the cache.

(should also add the requirements.txt and requirements-tests.txt hashes to the pyenv hash if installing these packages as part of the pyenv build)

Reply:
I added the the dependency to the `requirements-tests.txt` and pinned the version to `2.0.0`. As it also specified as a build-system dep with line
https://github.com/nod-ai/sharktank/blob/b1d09dbf1496949929c9aaf11f3cd9c551975958/libshortfin/pyproject.toml#L6
`nanobind` version `2.1.0` is installed when executing `pip install -v -e.` later. Should this be pinned to `2.0.0` as well?

Based on the work you've reviewed so far, I further improved caching. Instead of adding hashes for the requirements.txt files to the pyenv/python build step, I added an additional cache on top to cache the dependencies. Assuming that the dependencies are changing more rapidly, this will reduce the build time of the cache. The step to install and cache the dependencies takes about 60s whereas building pyenv/python takes ~ 5m 45s.